File: 03_javascript/01_two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
    const seen = new Map();

    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];

        if (seen.has(complement)) {
            return [seen.get(complement), i]
        }

        seen.set(nums[i], i);
    }

    return []; // unreachable as per problem guarantee
};

function runTest(nums, target) {
    const result = twoSum(nums, target);
    console.log(`nums=${JSON.stringify(nums)} target=${target} --> result=${JSON.stringify(result)}`);
}

// Sample LeetCode tests
runTest([2, 7, 11, 15], 9);
runTest([3, 2, 4], 6);
runTest([3, 3], 6);

File: 05_golang/01_two_sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	seen := make(map[int]int)

	for i, num := range nums {
		complement := target - num

		if idx, ok := seen[complement]; ok {
			return []int{idx, i}
		}

		seen[num] = i
	}

	return []int{} // unreachable as per problem
}

func runTest(nums []int, target int) {
	result := twoSum(nums, target)
	fmt.Printf("nums=%v target=%d --> result=%v\n", nums, target, result)
}

func main() {
	runTest([]int{2, 7, 11, 15}, 9)
	runTest([]int{3, 2, 4}, 6)
	runTest([]int{3, 3}, 6)
}
